core/src/main/java/org/elasticsearch/discovery/zen/ping/ZenPing.java

```diff
@@ -171,15 +193,15 @@ public synchronized boolean addPing(PingResponse ping) {
         }
 
         /** adds multiple pings if newer than previous pings from the same node */
-        public synchronized void addPings(PingResponse[] pings) {
+        public synchronized void addPings(Iterable<PingResponse> pings) {
             for (PingResponse ping : pings) {
                 addPing(ping);
             }
         }
 
-        /** serialize current pings to an array */
-        public synchronized PingResponse[] toArray() {
-            return pings.values().toArray(new PingResponse[pings.size()]);
+        /** serialize current pings to an array. It is guaranteed that the array contains one ping response per node */
+        public synchronized List<PingResponse> toList() {
+            return new ArrayList<>(pings.values());
```

Comment:
This is not a high performance code (especially not when this is called, where almost always things settled down) so I opted for the safest simplest option (imo) - return a new list and not worry what people do with it. Going with `CopyOnWriteHashMap` is a shame because it is updated quite frequently during the pinging phase.